File: python/graph_algo.py
from typing import List, Tuple


def dijkstra(n: int, graph: List[List[Tuple[int, int]]], src: int) -> List[int]:
    inf = 10**18
    dist = [inf] * n
    used = [False] * n
    dist[src] = 0

    for _ in range(n):
        u = -1
        for v in range(n):
            if not used[v] and (u == -1 or dist[v] < dist[u]):
                u = v
        if u == -1 or dist[u] == inf:
            break
        used[u] = True

        for to, w in graph[u]:
            if dist[to] > dist[u] + w:
                dist[to] = dist[u] + w

    return dist


def topological_sort(n: int, graph: List[List[int]]) -> List[int]:
    indeg = [0] * n
    for u in range(n):
        for v in graph[u]:
            indeg[v] += 1

    q = [i for i in range(n) if indeg[i] == 0]
    order = []
    i = 0
    while i < len(q):
        u = q[i]
        i += 1
        order.append(u)
        for v in graph[u]:
            indeg[v] -= 1
            if indeg[v] == 0:
                q.append(v)

    return order if len(order) == n else []

File: golang/graph/graph_algo.go
package graph

func Dijkstra(n int, g [][][2]int, src int) []int {
	const inf = int(1e18)
	dist := make([]int, n)
	used := make([]bool, n)
	for i := 0; i < n; i++ {
		dist[i] = inf
	}
	dist[src] = 0

	for i := 0; i < n; i++ {
		u := -1
		for v := 0; v < n; v++ {
			if !used[v] && (u == -1 || dist[v] < dist[u]) {
				u = v
			}
		}
		if u == -1 || dist[u] == inf {
			break
		}
		used[u] = true

		for _, e := range g[u] {
			to, w := e[0], e[1]
			if dist[to] > dist[u]+w {
				dist[to] = dist[u] + w
			}
		}
	}
	return dist
}

func TopologicalSort(n int, g [][]int) []int {
	indeg := make([]int, n)
	for u := 0; u < n; u++ {
		for _, v := range g[u] {
			indeg[v]++
		}
	}
	q := []int{}
	for i := 0; i < n; i++ {
		if indeg[i] == 0 {
			q = append(q, i)
		}
	}
	order := []int{}
	for i := 0; i < len(q); i++ {
		u := q[i]
		order = append(order, u)
		for _, v := range g[u] {
			indeg[v]--
			if indeg[v] == 0 {
				q = append(q, v)
			}
		}
	}
	if len(order) != n {
		return []int{}
	}
	return order
}
